##Sanity Check

In [1]:
import pandas as pd
import matplotlib.pyplot as plt

print("It worked!")



It worked!


##Load the data

In [2]:
df = pd.read_csv("data/spotify_dataset.csv", index_col=0) #index_col=0 gets rid of the Unnamed column in the dataset
df.head()


,track_id,artists,album_name,track_name,popularity,duration_ms,explicit,danceability,energy,key,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,time_signature,track_genre
0,5SuOikwiRyPMVoIQDJUgSV,Gen Hoshino,Comedy,Comedy,73,230666,False,0.676,0.4610,1,-6.746,0,0.1430,0.0322,0.000001,0.3580,0.715,87.917,4,acoustic
1,4qPNDBW1i3p13qLCt0Ki3A,Ben Woodward,Ghost (Acoustic),Ghost - Acoustic,55,149610,False,0.420,0.1660,1,-17.235,1,0.0763,0.9240,0.000006,0.1010,0.267,77.489,4,acoustic
2,1iJBSr7s7jYXzM8EGcbK5b,Ingrid Michaelson;ZAYN,To Begin Again,To Begin Again,57,210826,False,0.438,0.3590,0,-9.734,1,0.0557,0.2100,0.000000,0.1170,0.120,76.332,4,acoustic
3,6lfxq3CG4xtTiEg7opyCyx,Kina Grannis,Crazy Rich Asians (Original Motion Picture Sou...,Can't Help Falling In Love,71,201933,False,0.266,0.0596,0,-18.515,1,0.0363,0.9050,0.000071,0.1320,0.143,181.740,3,acoustic
4,5vjLSffimiIP26QG5WcN2K,Chord Overstreet,Hold On,Hold On,82,198853,False,0.618,0.4430,2,-9.681,1,0.0526,0.4690,0.000000,0.0829,0.167,119.949,4,acoustic


##Explore the dataset

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 114000 entries, 0 to 113999
Data columns (total 20 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   track_id          114000 non-null  object 
 1   artists           113999 non-null  object 
 2   album_name        113999 non-null  object 
 3   track_name        113999 non-null  object 
 4   popularity        114000 non-null  int64  
 5   duration_ms       114000 non-null  int64  
 6   explicit          114000 non-null  bool   
 7   danceability      114000 non-null  float64
 8   energy            114000 non-null  float64
 9   key               114000 non-null  int64  
 10  loudness          114000 non-null  float64
 11  mode              114000 non-null  int64  
 12  speechiness       114000 non-null  float64
 13  acousticness      114000 non-null  float64
 14  instrumentalness  114000 non-null  float64
 15  liveness          114000 non-null  float64
 16  valence           114000 

##Basic Statistics

In [4]:
df.describe()

,popularity,duration_ms,danceability,energy,key,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,time_signature
count,114000.000000,1.140000e+05,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000,114000.000000
mean,33.238535,2.280292e+05,0.566800,0.641383,5.309140,-8.258960,0.637553,0.084652,0.314910,0.156050,0.213553,0.474068,122.147837,3.904035
std,22.305078,1.072977e+05,0.173542,0.251529,3.559987,5.029337,0.480709,0.105732,0.332523,0.309555,0.190378,0.259261,29.978197,0.432621
min,0.000000,0.000000e+00,0.000000,0.000000,0.000000,-49.531000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,17.000000,1.740660e+05,0.456000,0.472000,2.000000,-10.013000,0.000000,0.035900,0.016900,0.000000,0.098000,0.260000,99.218750,4.000000
50%,35.000000,2.129060e+05,0.580000,0.685000,5.000000,-7.004000,1.000000,0.048900,0.169000,0.000042,0.132000,0.464000,122.017000,4.000000
75%,50.000000,2.615060e+05,0.695000,0.854000,8.000000,-5.003000,1.000000,0.084500,0.598000,0.049000,0.273000,0.683000,140.071000,4.000000
max,100.000000,5.237295e+06,0.985000,1.000000,11.000000,4.532000,1.000000,0.965000,0.996000,1.000000,1.000000,0.995000,243.372000,5.000000


Just from these basic statistics, it looks like most songs aren't very popular, with a mean of about 33 out of 100. That was probably the only surprising statistic. Other things to note are that songs average a duration of about 3.8 minutes, dancability seems to be very normally distributed, and most songs have very little "speechiness". It's also worth noting that features like Key and Mode are categorical features, represented by numeric values.

##**Questions**
1. What audio features have the biggest association with a song's popularity?
2. Are there distinct clusters (genres?) of songs based on audio features?
3. Can we predict if a song is popular?
4. Has song length or other features changed over time?

**##First, Clean the data.**
We'll start this by looking at the missing values to determine whether or not they are random to decide if we can drop them.

In [5]:
df.isna().sum().sort_values(ascending=False) #This gets the sum of null values in each colum in our dataframe

artists             1
album_name          1
track_name          1
track_id            0
popularity          0
duration_ms         0
explicit            0
danceability        0
energy              0
key                 0
loudness            0
mode                0
speechiness         0
acousticness        0
instrumentalness    0
liveness            0
valence             0
tempo               0
time_signature      0
track_genre         0
dtype: int64

It looks like there are only 3 null values in the entire dataset, and they have to do with the song name, album name, and artist. These should be fine to keep.

Next, let's check for duplicate values

In [6]:
df.duplicated().sum()

np.int64(450)

There were 450 duplicates.

**##Duplicate Removal**
Now, let's get clean the duplicate values to avoid bias towards frequently re-uploaded songs.

In [7]:
df = df.drop_duplicates(subset=['track_name', 'artists']) #Removes songs that have the same name and artist, so the same song on a different album, such as a top hit's album or deluxe version, aren't counted and don't bias the analysis.

**##Column Ranges**

In [8]:
df[['danceability','energy','speechiness','acousticness','valence']].describe() #Check

audio_features = [ #Features that should be between 0 - 1.
    'danceability','energy','speechiness',
    'acousticness','valence'
]

for col in audio_features: #loop through and remove rows where values are out of range
    df = df[(df[col] >= 0) & (df[col] <= 1)]

#popularity
df = df[(df['popularity'] >= 0) & (df['popularity'] <= 100)]


This enforces the range of values for each feature, removing invalid rows.

**##Fix Data Types**

In [9]:
df.dtypes #See types

df['duration_ms'] = pd.to_numeric(df['duration_ms'], errors='coerce') #Converts duration_ms to numeric value instead of scientific notation
df['popularity'] = df['popularity'].astype(int)

df = df.dropna() #Drop rows after conversion



**#Final check**

In [10]:
print("Final shape:", df.shape)
df.describe()


Final shape: (81343, 20)


,popularity,duration_ms,danceability,energy,key,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,time_signature
count,81343.000000,8.134300e+04,81343.000000,81343.000000,81343.000000,81343.000000,81343.000000,81343.000000,81343.000000,81343.000000,81343.000000,81343.000000,81343.000000,81343.000000
mean,34.636392,2.314145e+05,0.559276,0.635025,5.285900,-8.593930,0.632347,0.088993,0.329666,0.184733,0.219723,0.463277,122.144834,3.896967
std,19.438517,1.164924e+05,0.177747,0.258641,3.557629,5.304797,0.482169,0.116629,0.339961,0.331592,0.198272,0.263383,30.129012,0.456398
min,0.000000,8.586000e+03,0.000000,0.000000,0.000000,-49.531000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,21.000000,1.738710e+05,0.446000,0.455000,2.000000,-10.451500,0.000000,0.036100,0.015900,0.000000,0.098500,0.241000,99.408000,4.000000
50%,35.000000,2.152050e+05,0.573000,0.678000,5.000000,-7.262000,1.000000,0.049100,0.190000,0.000089,0.133000,0.449000,122.030000,4.000000
75%,49.000000,2.673460e+05,0.690000,0.857000,8.000000,-5.140000,1.000000,0.087000,0.629000,0.153000,0.283000,0.676000,140.128500,4.000000
max,100.000000,5.237295e+06,0.985000,1.000000,11.000000,4.532000,1.000000,0.965000,0.996000,1.000000,1.000000,0.995000,243.372000,5.000000


After cleaning, the dataset contains 81,343 songs with 20 features. I removed rows with incomplete features, removed duplicate tracks, validated feature ranges, and ensured correct data types. The final dataset is ready for analysis!